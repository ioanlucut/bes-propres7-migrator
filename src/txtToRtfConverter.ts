import fs from 'fs';
import path from 'path';

const RTF_TEMPLATE = 'rtf_mac_version_template.rtf';
const RTFS_DIR = 'rtfs';
const RTF_CONTENT_TEMPLATE = '$REPLACE_ME$';

let rtfTemplate: string;

const encodeVerseContentToUnicodeRTFFormat = (verseContent: string) =>
  verseContent
    // For some reason, if it's at the end of the word, we need to inject a new space char
    .replaceAll('Â ', '\\uc0\\u194  ')
    .replaceAll('â ', '\\uc0\\u226  ')
    .replaceAll('Ă ', '\\uc0\\u258  ')
    .replaceAll('ă ', '\\uc0\\u259  ')
    .replaceAll('Î ', '\\uc0\\u206  ')
    .replaceAll('î ', '\\uc0\\u238  ')
    .replaceAll('ș ', '\\uc0\\u537  ')
    .replaceAll('Ș ', '\\uc0\\u536  ')
    .replaceAll('ț ', '\\uc0\\u539  ')
    .replaceAll('Ț ', '\\uc0\\u538  ')
    .replaceAll('” ', '\\uc0\\u8221  ')
    .replaceAll('„ ', '\\uc0\\u8222  ')
    .replaceAll('’ ', '\\uc0\\u8217  ')
    // Remaining chars
    .replaceAll('Â', '\\uc0\\u194')
    .replaceAll('â', '\\uc0\\u226')
    .replaceAll('Ă', '\\uc0\\u258')
    .replaceAll('ă', '\\uc0\\u259')
    .replaceAll('Î', '\\uc0\\u206')
    .replaceAll('î', '\\uc0\\u238')
    .replaceAll('ș', '\\uc0\\u537')
    .replaceAll('Ș', '\\uc0\\u536')
    .replaceAll('ț', '\\uc0\\u539')
    .replaceAll('Ț', '\\uc0\\u538')
    .replaceAll('”', '\\uc0\\u8221')
    .replaceAll('„', '\\uc0\\u8222')
    .replaceAll('’', '\\uc0\\u8217');

export const convertToRtf = (multiLineVerseContent: string) => {
  if (!rtfTemplate) {
    rtfTemplate = fs
      .readFileSync(path.join(__dirname, '..', RTFS_DIR, RTF_TEMPLATE))
      .toString();
  }
  const normalizedContent = encodeVerseContentToUnicodeRTFFormat(
    multiLineVerseContent,
  ).replace(/\n/g, '\\\n');

  return rtfTemplate.replaceAll(RTF_CONTENT_TEMPLATE, normalizedContent);
};
