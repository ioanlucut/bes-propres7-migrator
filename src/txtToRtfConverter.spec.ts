import fs from 'fs';
import { convertToRtf } from './txtToRtfConverter';

describe('txtToRtfConverter', () => {
  it('should convert to expected `rtf` correctly', () => {
    const rtfContent = convertToRtf(`Row 1-1
Row 1-2 !(),-./1234567890:;?ABCDEFGHIJKLMNOPRSTUVWXZY[\\]abcdefghijklmnopqrstuvwxyzÎâîăÂȘșĂȚț’‘”„
Row 1-2 !(),-./1234567890:;?ABCDEFGHIJKLMNOPRSTUVWXZY[\\]abcdefghijklmnopqrstuvwxyzÎâîăÂȘșĂȚț’‘”„
Row 1-3`);

    expect(rtfContent).toMatchSnapshot();

    fs.writeFileSync('./txtToRtfConverterGeneratedRtf.rtf', rtfContent);
  });

  it('should convert to expected `rtf` correctly for edge cases', () => {
    const rtfContent = convertToRtf(`space-not-expected-before Îtext
space-not-expected-before âtext
space-not-expected-before îtext
space-not-expected-before ătext
space-not-expected-before Âtext
space-not-expected-before Ștext
space-not-expected-before ștext
space-not-expected-before Ătext
space-not-expected-before Țtext
space-not-expected-before țtext
space-not-expected-before -text
space-not-expected-before ’text
space-not-expected-before ‘text
space-not-expected-before ”text
space-not-expected-before „text

textÎ single-space-expected-after-this
textâ single-space-expected-after-this
textî single-space-expected-after-this
textă single-space-expected-after-this
textÂ single-space-expected-after-this
textȘ single-space-expected-after-this
textș single-space-expected-after-this
textĂ single-space-expected-after-this
textȚ single-space-expected-after-this
textț single-space-expected-after-this
text- single-space-expected-after-this
text’ single-space-expected-after-this
text‘ single-space-expected-after-this
text” single-space-expected-after-this
text„ single-space-expected-after-this

end-of-line-works-without-extra-space-textÎ
end-of-line-works-without-extra-space-textâ
end-of-line-works-without-extra-space-textî
end-of-line-works-without-extra-space-textă
end-of-line-works-without-extra-space-textÂ
end-of-line-works-without-extra-space-textȘ
end-of-line-works-without-extra-space-textș
end-of-line-works-without-extra-space-textĂ
end-of-line-works-without-extra-space-textȚ
end-of-line-works-without-extra-space-textț
end-of-line-works-without-extra-space-text-
end-of-line-works-without-extra-space-text’
end-of-line-works-without-extra-space-text‘
end-of-line-works-without-extra-space-text”
end-of-line-works-without-extra-space-text„`);

    expect(rtfContent).toMatchSnapshot();

    fs.writeFileSync('./txtToRtfConverterGeneratedRtf_II.rtf', rtfContent);
  });
});
