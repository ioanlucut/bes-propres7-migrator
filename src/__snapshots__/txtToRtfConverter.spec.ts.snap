// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`txtToRtfConverter should convert to expected \`rtf\` correctly 1`] = `
"{\\rtf1\\ansi\\ansicpg1252\\cocoartf2709\\cocoatextscaling0\\cocoaplatform0
{\\fonttbl\\f0\\fnil\\fcharset0 CMGSans-BoldCAPS;}
{\\colortbl;\\red255\\green255\\blue255;\\red255\\green255\\blue255;}
{\\*\\expandedcolortbl;;\\cssrgb\\c100000\\c100000\\c100000;}
\\paperw12240\\paperh15840\\vieww40760\\viewh23980\\viewkind0
\\pard\\tx560\\tx1120\\tx1680\\tx2240\\tx2800\\tx3360\\tx3920\\tx4480\\tx5040\\tx5600\\tx6160\\tx6720\\pardirnatural\\qc\\partightenfactor0

\\f0\\b\\fs140 \\cf2 \\cb3 Row 1-1\\
Row 1-2 !(),-./1234567890:;?ABCDEFGHIJKLMNOPRSTUVWXZY[\\]abcdefghijklmnopqrstuvwxyz\\uc0\\u206\\uc0\\u226\\uc0\\u238\\uc0\\u259\\uc0\\u194\\uc0\\u536\\uc0\\u537\\uc0\\u258\\uc0\\u538\\uc0\\u539\\uc0\\u8217\\uc0\\u8221\\uc0\\u8222\\
Row 1-3}
"
`;

exports[`txtToRtfConverter should convert to expected \`rtf\` correctly for edge cases 1`] = `
"{\\rtf1\\ansi\\ansicpg1252\\cocoartf2709\\cocoatextscaling0\\cocoaplatform0
{\\fonttbl\\f0\\fnil\\fcharset0 CMGSans-BoldCAPS;}
{\\colortbl;\\red255\\green255\\blue255;\\red255\\green255\\blue255;}
{\\*\\expandedcolortbl;;\\cssrgb\\c100000\\c100000\\c100000;}
\\paperw12240\\paperh15840\\vieww40760\\viewh23980\\viewkind0
\\pard\\tx560\\tx1120\\tx1680\\tx2240\\tx2800\\tx3360\\tx3920\\tx4480\\tx5040\\tx5600\\tx6160\\tx6720\\pardirnatural\\qc\\partightenfactor0

\\f0\\b\\fs140 \\cf2 \\cb3 space-not-expected-before \\uc0\\u206text\\
space-not-expected-before \\uc0\\u226text\\
space-not-expected-before \\uc0\\u238text\\
space-not-expected-before \\uc0\\u259text\\
space-not-expected-before \\uc0\\u194text\\
space-not-expected-before \\uc0\\u536text\\
space-not-expected-before \\uc0\\u537text\\
space-not-expected-before \\uc0\\u258text\\
space-not-expected-before \\uc0\\u538text\\
space-not-expected-before \\uc0\\u539text\\
space-not-expected-before -text\\
space-not-expected-before \\uc0\\u8217text\\
space-not-expected-before \\uc0\\u8221text\\
space-not-expected-before \\uc0\\u8222text\\
\\
text\\uc0\\u206  single-space-expected-after-this\\
text\\uc0\\u226  single-space-expected-after-this\\
text\\uc0\\u238  single-space-expected-after-this\\
text\\uc0\\u259  single-space-expected-after-this\\
text\\uc0\\u194  single-space-expected-after-this\\
text\\uc0\\u536  single-space-expected-after-this\\
text\\uc0\\u537  single-space-expected-after-this\\
text\\uc0\\u258  single-space-expected-after-this\\
text\\uc0\\u538  single-space-expected-after-this\\
text\\uc0\\u539  single-space-expected-after-this\\
text- single-space-expected-after-this\\
text\\uc0\\u8217  single-space-expected-after-this\\
text\\uc0\\u8221  single-space-expected-after-this\\
text\\uc0\\u8222  single-space-expected-after-this\\
\\
end-of-line-works-without-extra-space-text\\uc0\\u206\\
end-of-line-works-without-extra-space-text\\uc0\\u226\\
end-of-line-works-without-extra-space-text\\uc0\\u238\\
end-of-line-works-without-extra-space-text\\uc0\\u259\\
end-of-line-works-without-extra-space-text\\uc0\\u194\\
end-of-line-works-without-extra-space-text\\uc0\\u536\\
end-of-line-works-without-extra-space-text\\uc0\\u537\\
end-of-line-works-without-extra-space-text\\uc0\\u258\\
end-of-line-works-without-extra-space-text\\uc0\\u538\\
end-of-line-works-without-extra-space-text\\uc0\\u539\\
end-of-line-works-without-extra-space-text-\\
end-of-line-works-without-extra-space-text\\uc0\\u8217\\
end-of-line-works-without-extra-space-text\\uc0\\u8221\\
end-of-line-works-without-extra-space-text\\uc0\\u8222}
"
`;
